libraries/SITL/SIM_BalanceBot.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  Balance Bot simulator class
+*/
+
+#include "SIM_BalanceBot.h"
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+namespace SITL {
+
+BalanceBot::BalanceBot(const char *home_str, const char *frame_str) :
+    Aircraft(home_str, frame_str),
+    max_speed(4),
+    max_accel(14),
+    skid_turn_rate(140), // degrees/sec
+    skid_steering(true)
+{
+    dcm.from_euler(0,0,0); // initial yaw, pitch and roll in radians
+    printf("Balance Bot Simulation Started\n");
+}
+
+/*
+   return yaw rate in degrees/second given steering_angle
+*/
+float BalanceBot::calc_yaw_rate(float steering)
+{
+    return steering * skid_turn_rate;
+}
+
+/*
+  update the Balance Bot simulation by one time step
+ */
+/*
+ * WIP!
+ * The balance bot is physically modeled as an inverted pendulum(rod) on a cart
+ * Further details can be found here:
+ * 1) http://ctms.engin.umich.edu/CTMS/index.php?example=InvertedPendulum&section=SystemModeling
+ * 2) http://journals.sagepub.com/doi/pdf/10.5772/63933
+ */
+void BalanceBot::update(const struct sitl_input &input)
+{
+    const float length = 1.0f; //m length of body
+
+    const float mass_cart = 1.0f; // kg
+    const float mass_rod = 1.0f; //kg
+
+    // maximum force the motors can apply to the cart
+    const float max_force = 50.0f; //N
+
+    //Moment of Inertia of the rod
+    const float I_rod = (mass_rod*4*length*length)/12.0f; //kg.m^2
+
+    // air resistance
+    const float damping_constant = 0.7; // N-s/m
+
+    float steering,throttle;
+
+    // balance bot uses skid steering
+    float motor1 = 2*((input.servos[0]-1000)/1000.0f - 0.5f);
+    float motor2 = 2*((input.servos[2]-1000)/1000.0f - 0.5f);
+    steering = motor1 - motor2;
```

Comment:
const float steering = motor1 - motor2;
const float throttle = 0.5f * (motor1 + motor2);

For the file, try to const the most and do direct initialisation of variable: that c++ not C ! ;-)